File: wehelp/week2/assignment.py
print("=== Task1 ===")
"""
Task 1:
We have example messages from 6 persons in JSON format. There are at least 3 persons
who are older than 17. Find out those who are most probably older than 17 years old based
on example messages. Print their names in the console.
"""

def find_and_print(messages):
    # write down your judgment rules in comments
    # your code here, based on your own rules
    keywords = ["18 years old", "college student", "legal age", "vote"]
    for name, v in messages.items():
        for condition in keywords:
            if condition in v:
                print(name)

find_and_print({
"Bob":"My name is Bob. I'm 18 years old.",
"Mary":"Hello, glad to meet you.",
"Copper":"I'm a college student. Nice to meet you.",
"Leslie":"I am of legal age in Taiwan.",
"Vivian":"I will vote for Donald Trump next week",
"Jenny":"Good morning."
})

print("=== Task2 ===")
"""
Task 2:
Complete functions below to calculate sum of bonus of all employees in TWD and print it.
1. Bonus should depend on salary, performance and role fields. Define your own rules
and calculate a bonus for each employee based on it.
2. Sum of bonus of all employees cannot exceed 10000 TWD based on your rules and
example data.
3. You can assume the USD to TWD Exchange Rate is 30.
4. Salary is default to TWD if there is no specific mark.
"""
def calculate_sum_of_bonus(data):
    """
    performance bonus ratio "above average" : 5%
    performance bonus ratio "average": 3%
    performance bonus ratio "below average": 1%

    role bonus ratio "Engineer": 5%
    role bonus ratio "Sales": 3%
    role bonus ratio "CEO": 1%

    The bonus of each staff is: min(salary * (performance ratio + role ratio), 2500)
    """
    # write down your bonus rules in comments
    # your code here, based on your own rules

    total_bonus = 0
    for staff in data["employees"]:
        staff['bonus_ratio'] = 0

        if staff["performance"] == "above average":
            staff['bonus_ratio'] += 0.05
        elif staff["performance"] == "average":
            staff['bonus_ratio'] += 0.03
        elif staff["performance"] == "below average":
            staff['bonus_ratio'] += 0.01

        if staff["role"] == "Engineer":
            staff['bonus_ratio'] += 0.05
        elif staff["role"] == "Sales":
            staff['bonus_ratio'] += 0.03
        elif staff["role"] == "CEO":
            staff['bonus_ratio'] += 0.01

        if isinstance(staff["salary"], str):
            if "USD" in staff["salary"]:
                staff["salary"] = int(staff["salary"].replace("USD", "")) * 30
            elif "," in staff["salary"]:
                staff["salary"] = int(staff["salary"].replace(",", ""))

        staff["bonus"] = min(staff["salary"] * staff['bonus_ratio'], 2500)
        
        total_bonus += staff["bonus"]

    print(f"total bonus: {total_bonus}")

calculate_sum_of_bonus({
"employees":[
{
"name":"John",
"salary":"1000USD",
"performance":"above average",
"role":"Engineer"
},
{
"name":"Bob",
"salary":60000,
"performance":"average",
"role":"CEO"
},
{
"name":"Jenny",
"salary":"50,000",
"performance":"below average",
"role":"Sales"
}
]
}) # call calculate_sum_of_bonus function

print("=== Task3 ===")
"""
Task 3:
Find out whose middle name is unique among all the names, and print it. You can assume
every input is a Chinese name with 2 ~ 3 words. If there are only 2 words in a name, the
middle name is defined as the second word.
"""

def func(*data):
    # your code here
    temp = {}
    for name in data:
        temp[name[1]] = temp.get(name[1], [])
        temp[name[1]].append(name)

    for k, v in temp.items():
        if len(v) == 1:
            print(v[0])
            return
    
    print("沒有")
    return

func("彭大牆", "王明雅", "吳明") # print 彭大牆
func("郭靜雅", "王立強", "林靜宜", "郭立恆", "林花花") # print 林花花
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花") # print 沒有

print("=== Task4 ===")
"""
Task 4:
There is a number sequence: 0, 4, 3, 7, 6, 10, 9, 13, 12, 16, 15, ...
Find out the nth term in this sequence.
"""

def get_number(index:int):
    if index >= 0:
        ret = ((index + 1) // 2 * 4) - ((index // 2))
        print(f"Ans: {ret}")

get_number(1) # print 4
get_number(5) # print 10
get_number(10) # print 15

print("=== Task5 ===")
"""
Task 5 (Optional):
Given available seats for each car of a train, status bitmap, and number of incoming
passengers, writing a procedure to find out the index of the most fitted car to serve
passengers. Print -1 if there is no car which can serve incoming passengers.
- Available Seats: list/array containing number of available seats for each car.
- Status Bitmap: list/array containing only 0 or 1. 1 means the corresponding car can
serve passengers for now.
- Passenger Number: number of incoming passengers.
We can assume all incoming passengers should be served in the same car.
"""

def find_index_of_car(seats, status, number):
    # your code here
    ret = {}
    for i, v in enumerate(status):
        if v == 1:
            if seats[i] >= number:
                ret[i] = seats[i] - number
    if not ret:
        min_key = -1
    else:
        min_key = min(ret, key=ret.get)
    print(min_key)

find_index_of_car([3, 1, 5, 4, 2], [0, 1, 0, 1, 1], 2) # print 4
find_index_of_car([1, 0, 5, 1, 3], [0, 1, 0, 1, 1], 4) # print -1
find_index_of_car([4, 6, 5, 8], [0, 1, 1, 1], 4) # print 2

File: wehelp/week2/assignment.js
console.log("=== Task 1 ===")
function findAndPrint(messages){
    // write down your judgment rules in comments
    // your code here, based on your own rules
    const keywords = ["18 years old", "college student", "legal age", "vote"];
    for(var name in messages){
        for(var key in keywords){
            // console.log(keywords[key])
            if(messages[name].includes(keywords[key])){
                console.log(name);
            }
        }
    }
}
findAndPrint({
"Bob":"My name is Bob. I'm 18 years old.",
"Mary":"Hello, glad to meet you.",
"Copper":"I'm a college student. Nice to meet you.",
"Leslie":"I am of legal age in Taiwan.",
"Vivian":"I will vote for Donald Trump next week",
"Jenny":"Good morning."
});

console.log("=== Task 2 ===")
function calculateSumOfBonus(data){
    // write down your bonus rule in comments
    // your code here, based on your own rules

    // bonus rule:
    // performance bonus ratio "above average" : 5%
    // performance bonus ratio "average": 3%
    // performance bonus ratio "below average": 1%

    // role bonus ratio "Engineer": 5%
    // role bonus ratio "Sales": 3%
    // role bonus ratio "CEO": 1%

    // The bonus of each staff is: Math.min(salary * (performance ratio + role ratio), 2500)
    var total_bonus = 0;
    for(i in data["employees"]){
        var salary = data["employees"][i]["salary"];
        var performance = data["employees"][i]["performance"];
        var role = data["employees"][i]["role"];
        var bonus_ratio = 0;

        if (performance == "above average") {
            bonus_ratio += 0.05;
        } else if (performance == "average") {
            bonus_ratio += 0.03;
        } else if (performance == "below average") {
            bonus_ratio += 0.01;
        }
        
        if (role == "Engineer") {
            bonus_ratio += 0.05;
        } else if (role == "Sales") {
            bonus_ratio += 0.03;
        } else if (role == "CEO") {
            bonus_ratio += 0.01;
        }

        if (typeof salary === "string"){
            if (salary.includes("USD")){
                data["employees"][i]["salary"] = parseInt(salary.replace(/USD/, "")) * 30;
            }else if (salary.includes(",")){
                data["employees"][i]["salary"] = parseInt(salary.replace(/,/, ""));
            }
        }

        data["employees"][i]['bonus'] = Math.min(data["employees"][i]["salary"] * bonus_ratio, 2500);
        
        total_bonus += data["employees"][i]['bonus'];
    }
    console.log(total_bonus);
}
calculateSumOfBonus({
"employees":[
{
"name":"John",
"salary":"1000USD",
"performance":"above average",
"role":"Engineer"
},
{
"name":"Bob",
"salary":60000,
"performance":"average",
"role":"CEO"
},
{
"name":"Jenny",
"salary":"50,000",
"performance":"below average",
"role":"Sales"
}
]
}); // call calculateSumOfBonus function

console.log("=== Task 3 ===")
function func(...data){
    // your code here
    var temp = {};
    data.forEach(function(name) {
        temp[name[1]] = temp[name[1]] || [];
        temp[name[1]].push(name);
    });
    for (i in temp){
        if (temp[i].length == 1){
            console.log(temp[i][0]);
            return;
        }
    }
    console.log("沒有");
}
func("彭大牆", "王明雅", "吳明"); // print 彭大牆
func("郭靜雅", "王立強", "林靜宜", "郭立恆", "林花花"); // print 林花花
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花"); // print 沒有

console.log("=== Task 4 ===")
function getNumber(index){
    // your code here
    if (index >= 0){
        var ret = (Math.floor((index + 1) / 2) * 4) - Math.floor(index / 2);
        console.log(ret);
    }
}
getNumber(1); // print 4
getNumber(5); // print 10
getNumber(10); // print 15

console.log("=== Task 5 ===")
function findIndexOfCar(seats, status, number){
    // your code here
    var ret = {};
    var car = null;
    for (i in status){
        if (status[i] == 1){
            if (seats[i] >= number){
                ret[i] = seats[i] - number;
            }
        }
    }
    if (Object.keys(ret).length === 0) {
        car = -1;
    } else {
        car = Object.keys(ret).reduce(function(a, b) {
            return ret[a] < ret[b] ? a : b;
        });
    }
    console.log(car)
}
findIndexOfCar([3, 1, 5, 4, 2], [0, 1, 0, 1, 1], 2); // print 4
findIndexOfCar([1, 0, 5, 1, 3], [0, 1, 0, 1, 1], 4); // print -1
findIndexOfCar([4, 6, 5, 8], [0, 1, 1, 1], 4); // print 2